src/main/scala/com/cognite/sdk/scala/v1/resources/dataModels.scala

```diff
@@ -60,28 +56,137 @@ class DataModels[F[_]](val requestSession: RequestSession[F])
+        case spaceExternalId :: modelExternalId :: Nil =>
+          DataModelIdentifier(Some(spaceExternalId), modelExternalId)
+        case list =>
+          throw new SdkException(
+            s"Unable to decode DataModelIdentifier, expected array length 1 or 2, actual length ${list.length.toString}."
+          )
+      }
+  implicit val bTreeIndexDecoder: Decoder[BTreeIndex] =
+    deriveDecoder[BTreeIndex]
+  implicit val dataModelPropertyTypeDecoder: Decoder[PropertyType] =
+    Decoder.decodeString.map(
+      PropertyType
+        .fromCode(_)
+        .getOrElse(
+          throw new IllegalArgumentException("Invalid type specified")
+        )
+    )
+  implicit val dataModelPropertyIndexesDecoder: Decoder[DataModelIndexes] =
+    deriveDecoder[DataModelIndexes]
   implicit val dataModelPropertyDecoder: Decoder[DataModelProperty] =
     deriveDecoder[DataModelProperty]
-  implicit val dataModelDecoder: Decoder[DataModel] = deriveDecoder[DataModel]
-  implicit val dataModelItemsDecoder: Decoder[Items[DataModel]] = deriveDecoder[Items[DataModel]]
-
-  implicit val dataModelGetByCogniteIdIdsEncoder
-      : Encoder[DataModelGetByExternalIdsInput[CogniteId]] =
-    deriveEncoder[DataModelGetByExternalIdsInput[CogniteId]]
-
-  implicit val dataModelGetByExternalIdsEncoder
-      : Encoder[DataModelGetByExternalIdsInput[CogniteExternalId]] =
-    deriveEncoder[DataModelGetByExternalIdsInput[CogniteExternalId]]
+  implicit val uniquenessConstraintDecoder: Decoder[UniquenessConstraint] =
+    deriveDecoder[UniquenessConstraint].prepare(init =>
+      init
+        .downField("uniqueProperties")
+        .set(
+          Json.fromValues(
+            init
+              .downField("uniqueProperties")
+              .values
+              .getOrElse(throw new Exception("uniqueProperties is required"))
+              .map(v =>
+                v.hcursor
+                  .downField("property")
+                  .focus
+                  .getOrElse(throw new Exception("property is required"))
+              )
+              .toVector
+          )
+        )
+    )
+  implicit val dataModelConstraintsDecoder: Decoder[DataModelConstraints] =
+    deriveDecoder[DataModelConstraints]
+  implicit val dataModelInstanceTypeDecoder: Decoder[DataModelInstanceType] =
+    deriveDecoder[DataModelInstanceType]
+  implicit val dataModelDecoder: Decoder[DataModel] =
+    deriveDecoder[DataModel].prepare { init =>
+      val instanceType: Json =
+        dataModelInstanceTypeEncoder(
+          init
+            .as[DataModelInstanceType]
+            .fold(
+              _ => throw new Exception("allowNode and allowEdge are required"),
+              t => t
+            )
+        )
 
+      init.withFocus(json =>
+        Json.fromJsonObject(
+          json.asObject
+            .getOrElse(throw new Exception("invalid json object"))
+            .+:("instanceType" -> instanceType)
+        )
+      )
+    }
```

Comment:
Didn't quite get this part, basically we try casting the entire thing to a `DataModelInstanceType` first, to check for allow edge/node, and in the second part we check it's valid Json?

We don't usually do validation on the CDF API stuff, we just assume it works (reading from the CDF API is the only case where we need to decode JSON).. as for nested case classes like DataModel, if there's special handling needed it's generally better to rely on separate decoders for the lower case classes